src/components/Tables/Table.tsx

```diff
@@ -41,86 +40,59 @@ export interface ColumnStyleOptions {
+    defaultShownHeaders?: readonly (keyof Data)[];
+    toggleableHeaders?: readonly (keyof Data)[];
+    sortable?: (keyof Data)[];
     onEdit?: (data: number) => void;
     onDelete?: (data: number) => void;
-    columnDisplayFunctions?: TableColumnDisplayFunctions;
-    columnStyleOptions?: TableColumnStyleOptions;
-    onRowClick?: OnRowClickFunction;
+    columnDisplayFunctions?: { [headerKey in keyof Data]?: ColumnDisplayFunction<Data[headerKey]> };
+    columnStyleOptions?: { [headerKey in keyof Data]?: ColumnStyleOptions };
+    onRowClick?: OnRowClickFunction<Data>;
+    autoFilter?: boolean;
 }
 
-const doesRowIncludeFilterText = (
-    row: Row,
-    filterText: FilterText,
-    headers: TableHeaders
-): boolean => {
-    for (const [headerKey, _headerLabel] of headers) {
-        if (
-            !(row.data[headerKey] ?? "")
-                .toString()
-                .toLowerCase()
-                .includes((filterText[headerKey] ?? "").toLowerCase())
-        ) {
-            return false;
-        }
-    }
-    return true;
-};
-
-const dataToFilteredRows = (
-    data: Datum[],
-    filterText: FilterText,
-    headers: TableHeaders
-): Row[] => {
-    const rows = dataToRows(data, headers);
-    return filterRows(rows, filterText, headers);
-};
-
-const dataToRows = (data: Datum[], headers: TableHeaders): Row[] => {
-    return data.map((datum: Datum, currentIndex: number) => {
-        const row: Row = { rowId: currentIndex, data: { ...datum } };
-
-        for (const [headerKey, _headerLabel] of headers) {
-            const databaseValue = datum[headerKey] ?? "";
-            row.data[headerKey] = Array.isArray(databaseValue)
-                ? databaseValue.join(", ")
-                : databaseValue;
-        }
-
-        return row;
-    });
-};
-
-const filterRows = (rows: Row[], filterText: FilterText, headers: TableHeaders): Row[] => {
-    return rows.filter((row) => doesRowIncludeFilterText(row, filterText, headers));
-};
-
-interface CellProps {
-    row: Row;
-    columnDisplayFunctions: TableColumnDisplayFunctions;
-    headerKey: string;
+interface CellProps<Data> {
+    row: Row<Data>;
+    columnDisplayFunctions: { [headerKey in keyof Data]?: ColumnDisplayFunction<Data[headerKey]> };
+    headerKey: keyof Data;
 }
 
-const CustomCell: React.FC<CellProps> = ({ row, columnDisplayFunctions, headerKey }) => {
-    return (
+const CustomCell = <Data extends unknown>({
+    row,
+    columnDisplayFunctions,
+    headerKey,
+}: CellProps<Data>): React.ReactElement => {
+    const element = (
         <>
             {columnDisplayFunctions[headerKey]
-                ? columnDisplayFunctions[headerKey](row)
+                ? columnDisplayFunctions[headerKey]!(row.data[headerKey])
                 : row.data[headerKey]}
         </>
     );
+
+    if (!React.isValidElement(element)) {
+        throw new Error(
+            `${element} is not a valid JSX element, add a column display function for ${String(
```

Comment:
(genuinely) not sure if the error message is consistent with VFB-22 so maybe add a TODO VFB-22 if yours is merged before that, or change it (if necessary)?

Reply:
This shouldn't ever be displayed to the user, this is a broken API invariant. This could probably be fixed with more exotic types but that would come at the expense of code readability so this is what I thought the best solution would be